src/accelerate/utils/operations.py

```diff
@@ -117,13 +117,13 @@ def send_to_device(tensor, device):
         The same data structure as `tensor` with all tensors sent to the proper device.
     """
 
-    def _send_to_device(t, device):
-        return t.to(device)
+    def _send_to_device(t, device, non_blocking):
+        return t.to(device, non_blocking=non_blocking)
 
     def _has_to_method(t):
         return hasattr(t, "to")
```

Comment:
So far all objects in Pytorch which accept `.to()` method accept the `non_blocking` kwarg or at least kwargs in general, but are we sure this is going to be always the case?

- [torch.jit.ScriptModule.to](https://pytorch.org/docs/stable/generated/torch.jit.ScriptModule.html?highlight=#torch.jit.ScriptModule.to) (Python method, in ScriptModule)
- [torch.nn.Module.to](https://pytorch.org/docs/stable/generated/torch.nn.Module.html?highlight=#torch.nn.Module.to) (Python method, in Module)
- [torch.nn.utils.rnn.PackedSequence.to](https://pytorch.org/docs/stable/generated/torch.nn.utils.rnn.PackedSequence.html?highlight=#torch.nn.utils.rnn.PackedSequence.to) (Python method, in PackedSequence)
- [torch.Tensor.to](https://pytorch.org/docs/stable/generated/torch.Tensor.to.html?highlight=#torch.Tensor.to) (Python method, in torch.Tensor.to)

Reply:
This is for our own objects FYI, `BatchEncoding` in Transformers, which does not accept the `non_blocking` argument :-/
Maybe we can inspect the signature and only pass it when we see it?

Reply:
Mmm, `inspect.signature(t.to)` should give you access to the keyword arguments. You can find them in the `.parameters` attribute, which will be a dict-like object.

Reply:
I get this error when trying `inspect.signature(t.to)` where `t` is a tensor
```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/nouamane/miniconda3/envs/hf/lib/python3.9/inspect.py", line 3111, in signature
    return Signature.from_callable(obj, follow_wrapped=follow_wrapped)
  File "/home/nouamane/miniconda3/envs/hf/lib/python3.9/inspect.py", line 2860, in from_callable
    return _signature_from_callable(obj, sigcls=cls,
  File "/home/nouamane/miniconda3/envs/hf/lib/python3.9/inspect.py", line 2327, in _signature_from_callable
    return _signature_from_builtin(sigcls, obj,
  File "/home/nouamane/miniconda3/envs/hf/lib/python3.9/inspect.py", line 2145, in _signature_from_builtin
    raise ValueError("no signature found for builtin {!r}".format(func))
ValueError: no signature found for builtin <built-in method to of Tensor object at 0x7f7e98c57720>
```

Reply:
I think this is related to the warning in [inspect's docs](https://docs.python.org/3/library/inspect.html#inspect.signature)

> Note: Some callables may not be introspectable in certain implementations of Python. For example, in CPython, some built-in functions defined in C provide no metadata about their arguments.

Reply:
Maybe we can isolate tensors in the function then? Something like:
```
if isinstance(t, torch.Tensor):
    return t.to(device, non_blocking=non_blocking)
else:
    # inspect signature here in a try/except block
```

Reply:
Wouldn't this be enough, or should I inspect signature?
  ```python
  def _send_to_device(t, device, non_blocking):
        if isinstance(t, torch.Tensor):
            return t.to(device, non_blocking=non_blocking)
        else:
            try:
                return t.to(device, non_blocking=non_blocking)
            except TypeError:
                return t.to(device)
```

Reply:
Oh in this case, you don't even need the first test. Are you sure about the type of `TypeError`?

Reply:
Seems to work for C defined methods as well:
```python
>> t.__abs__(a=5)
Traceback (most recent call last):
  File "<string>", line 1, in <module>
TypeError: Tensor.abs() takes no keyword arguments
```


```python
>> t.to(a=5)
Traceback (most recent call last):
  File "<string>", line 1, in <module>
TypeError: to() received an invalid combination of arguments - got (a=int, ), but expected one of:
 * (torch.device device, torch.dtype dtype, bool non_blocking, bool copy, *, torch.memory_format memory_format)
 * (torch.dtype dtype, bool non_blocking, bool copy, *, torch.memory_format memory_format)
 * (Tensor tensor, bool non_blocking, bool copy, *, torch.memory_format memory_format)
```

Reply:
Nice!